src/main/java/com/google/api/generator/gapic/protoparser/HttpRuleParser.java

```diff
@@ -53,26 +54,23 @@
     return parseHttpRuleHelper(httpRule, Optional.of(inputMessage), messageTypes);
   }
 
-  public static Optional<List<String>> parseHttpRule(HttpRule httpRule) {
+  public static HttpRuleBindings parseHttpRule(HttpRule httpRule) {
     return parseHttpRuleHelper(httpRule, Optional.empty(), Collections.emptyMap());
   }
 
-  private static Optional<List<String>> parseHttpRuleHelper(
+  private static HttpRuleBindings parseHttpRuleHelper(
       HttpRule httpRule, Optional<Message> inputMessageOpt, Map<String, Message> messageTypes) {
     // Get pattern.
-    Set<String> uniqueBindings = getHttpVerbPattern(httpRule);
-    if (uniqueBindings.isEmpty()) {
-      return Optional.empty();
-    }
-
+    String pattern = getPattern(httpRule);
+    ImmutableSet.Builder<String> bindingsBuilder = getPatternBindings(pattern);
     if (httpRule.getAdditionalBindingsCount() > 0) {
       for (HttpRule additionalRule : httpRule.getAdditionalBindingsList()) {
-        uniqueBindings.addAll(getHttpVerbPattern(additionalRule));
+        // TODO: save additional bindings path in HttpRuleBindings
```

Comment:
Is this TODO still needed?

Reply:
Yes, as it talks about saving the `path`. Though I'm not sure if it would be needed at all.

Reply:
Should we add this TODO to HttpRuleBindings as well?